Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2009, 2015, 2019, 2020, 2024 Igalia S.L.
+ * Copyright (C) 2015, 2019 Metrological Group B.V.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CoordinatedPlatformLayerBufferVideo.h"
+
+#if USE(COORDINATED_GRAPHICS) && ENABLE(VIDEO) && USE(GSTREAMER)
+#include "CoordinatedPlatformLayerBufferExternalOES.h"
+#include "CoordinatedPlatformLayerBufferRGB.h"
+#include "CoordinatedPlatformLayerBufferYUV.h"
+#include "TextureMapper.h"
+
+// Include the <epoxy/gl.h> header before <gst/gl/gl.h>.
+#include <epoxy/gl.h>
+#include <gst/gl/gl.h>
+
+namespace WebCore {
+
+std::unique_ptr<CoordinatedPlatformLayerBufferVideo> CoordinatedPlatformLayerBufferVideo::create(GstSample* sample, std::optional<GstVideoDecoderPlatform> videoDecoderPlatform, bool gstGLEnabled, OptionSet<TextureMapperFlags> flags)
```

Comment:
perhaps use `const GRefPtr<GstSample>&` here and keep it as m_sample?

Also you could use `GstMappedFrame` also, instead of the more manual work of calling the gst_video_frame_map and friends.

Reply:
> perhaps use `const GRefPtr<GstSample>&` here and keep it as m_sample?

Do we need to keep a reference of the sample?

> Also you could use `GstMappedFrame` also, instead of the more manual work of calling the gst_video_frame_map and friends.

I thought about it, but decided to just try to copy existing code as much as possible in this patch.

Reply:
> Do we need to keep a reference of the sample?

It might be safer yes. The buffer we get from it has a lifetime bound to it. Or we could also just keep a ref of the buffer.

Reply:
We are already keeping a ref of the buffer